azure/services/subnets/spec.go

```diff
@@ -43,107 +45,71 @@ type SubnetSpec struct {
 	ServiceEndpoints  infrav1.ServiceEndpoints
 }
 
-// ResourceName returns the name of the subnet.
-func (s *SubnetSpec) ResourceName() string {
-	return s.Name
-}
-
-// ResourceGroupName returns the name of the resource group of the VNet that owns this subnet.
-func (s *SubnetSpec) ResourceGroupName() string {
-	return s.VNetResourceGroup
-}
-
-// OwnerResourceName returns the name of the VNet that owns this subnet.
-func (s *SubnetSpec) OwnerResourceName() string {
-	return s.VNetName
+// ResourceRef implements azure.ASOResourceSpecGetter.
+func (s *SubnetSpec) ResourceRef() *asonetworkv1.VirtualNetworksSubnet {
+	return &asonetworkv1.VirtualNetworksSubnet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.VNetName + "-" + strings.ToLower(s.Name),
+			Namespace: s.Namespace,
+		},
+	}
 }
 
-// Parameters returns the parameters for the subnet.
-func (s *SubnetSpec) Parameters(ctx context.Context, existing interface{}) (parameters interface{}, err error) {
-	if existing != nil {
-		existingSubnet, ok := existing.(armnetwork.Subnet)
-		if !ok {
-			return nil, errors.Errorf("%T is not an armnetwork.Subnet", existing)
-		}
-
-		if !s.shouldUpdate(existingSubnet) {
-			return nil, nil
-		}
+// Parameters implements azure.ASOResourceSpecGetter.
+func (s *SubnetSpec) Parameters(ctx context.Context, existing *asonetworkv1.VirtualNetworksSubnet) (parameters *asonetworkv1.VirtualNetworksSubnet, err error) {
+	subnet := existing
+	if subnet == nil {
+		subnet = &asonetworkv1.VirtualNetworksSubnet{}
 	}
 
-	if !s.IsVNetManaged {
-		// TODO: change this to terminal error once we add support for handling them
-		return nil, errors.Errorf("custom vnet was provided but subnet %s is missing", s.Name)
```

Comment:
This check is currently being removed and I'll edit this comment to explain why later.

Reply:
I think that was actually a requested user feature in the past (there might even be an issue for it somewhere)

How do we handle delete for those vnets/subnets?

Reply:
My archaeology skills led me here, where it seems this was added as a part of a general effort to support BYO vnet: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/340/files#diff-6d4bea1a244164bf3d16b025868b878dffcf0c78963b59bad9c189bcee04d35eR87-R90

CAPZ would take full responsibility for any subnet it creates, whether it's part of a managed or unmanaged vnet. So a managed subnet in an unmanaged vnet can be deleted by CAPZ.

Reply:
^ But in that same scenario, the unmanaged vnet would not be deleted.